### PR TITLE
http: support route mutability

### PR DESCRIPTION
**Pull request by mengmichael1**
Commit Message: http: support route mutability

Additional Description: Implements https://github.com/envoyproxy/envoy/issues/14904. 
- Adds a `setRoute` method into the Filter API, adding `setRoute` into HCM `ActiveStream` where it is implemented as a `StreamFilterCallbacks` for filters to call
- Adds a `DelegatingRoute` / `DelegatingRouteEntry` mechanism to easily create a route that mutates specific properties of an existing route.

Risk Level: Low, implemented as new feature (no behavior changes)

Testing: 
- Unit tests in `HttpConnectionManagerImplTest`: testing (1) happy path case of setting the route to a DelegatingRoute that overrides clusterName, (2) setRoute(nullptr), (3) all routeEntry calls are delegated correctly and exhibit the same behavior as the base route's routeEntry
- Integration test with `(IntegrationTest, SetRouteToDelegatingRouteWithClusterOverride)`

Docs Changes: "Life of a Request" docs, HTTP filters arch docs

Release Notes: Added to 1.18.0
``` 
http: added support for stream filters to mutate the cached route set by HCM route resolution. Useful for filters in a filter chain that want to override specific methods/properties of a route. See :ref:`http route mutation <arch_overview_http_filters_route_mutation>` docs for more information.
```

Platform Specific Features: n/a

[Optional [API Considerations](https://github.com/envoyproxy/envoy/blob/main/api/review_checklist.md):] Should be noted that all changes (e.g. addition of a new virtual method) to the `Route` / `RouteEntry` API (in `include/envoy/router`) must also be changed in the `DelegatingRoute` / `DelegatingRouteEntry` API. Developers will see build errors if there are new virtual methods not implemented, which will act as a forcing function to maintain parity b/w the APIs.

### CODE REVIEW

File: docs/root/intro/life_of_a_request.rst
Comment (snowp) [inline docs/root/intro/life_of_a_request.rst:437]: wrap lines at 100 characters

File: include/envoy/http/filter.h
Comment (snowp) [inline include/envoy/http/filter.h:223]: nit: don't use `const` for value parameters, just do `Router::RouteConstSharedPtr`

File: source/common/http/filter_manager.h
Comment (snowp) [inline source/common/http/filter_manager.h:629]: same, remove `const`

File: include/envoy/http/filter.h
Comment (snowp) [inline include/envoy/http/filter.h:223]: What happens if we pass nullptr here? Does that make it equivalent to clearRouteCache? Should we consider disallowing nullptr here?

File: source/common/router/delegating_route_impl.h
Comment (snowp) [inline source/common/router/delegating_route_impl.h:18]: make single arg ctors `explicit`

File: source/common/router/delegating_route_impl.h
Comment (snowp) [inline source/common/router/delegating_route_impl.h:32]: The code in the .cc file assumes that this is nullptr, so passing `const Router::RouteEntry&` seems better.

We should probably document the lifetime guarantees and perhaps talk about why its fine for this class to take the delegated object as a non-owning ref while the `DelegatingRoute` takes a shared_ptr

File: source/common/router/delegating_route_impl.h
Comment (snowp) [inline source/common/router/delegating_route_impl.h:30]: doc string

File: source/common/router/delegating_route_impl.h
Comment (snowp) [inline source/common/router/delegating_route_impl.h:38]: usually we just comment with the name of the base class we're implementing functions from, so `// Router::RouteEntry`

File: test/common/http/conn_manager_impl_test_base.h
Comment (snowp) [inline test/common/http/conn_manager_impl_test_base.h:234]: make `custom_route_entry` a unique_ptr and do `std::make_unique< ExampleDelegatingRouteEntryDerived>(base_route->routeEntry(), cluster_name_override)` to avoid a memory leak

File: test/integration/protocol_integration_test.cc
Comment (snowp) [inline test/integration/protocol_integration_test.cc:142]: Is using original dst cluster necessary here? Seems like it complicates the test

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks this looks like the right approach! Some comments to get you started, I didn't dig into the tests that much

File: docs/root/intro/life_of_a_request.rst
Comment (mengmichael1) [reply docs/root/intro/life_of_a_request.rst:437]: Is there a linter / script somewhere I can run to easily format rst? I tried building docs per https://github.com/envoyproxy/envoy/blob/main/docs/README.md but that didn't do anything

File: source/common/router/delegating_route_impl.h
Comment (mengmichael1) [reply source/common/router/delegating_route_impl.h:32]: 1. Hmm sorry what do you mean by "assumes that this is nullptr"? I had it using `const Router::RouteEntry*` to align with `const Router::RouteEntry* routeEntry() const override;` from the `DelegatingRoute` class.

2. Agree that we should document lifetime guarantees and ownership choices made. Let me think about this for a bit. What did you have in mind? I chose `DelegatingRoute` to take a shared_ptr to align with existing route functionality/methods such as [route() returning a RouteConstSharedPtr ](https://github.com/envoyproxy/envoy/blob/v1.17.1/include/envoy/http/filter.h#L197)

File: docs/root/intro/life_of_a_request.rst
Comment (snowp) [reply docs/root/intro/life_of_a_request.rst:437]: I use `tools/code_format/check_format.py fix` to format but I'm not sure if it covers the RST files

File: source/common/router/delegating_route_impl.h
Comment (snowp) [reply source/common/router/delegating_route_impl.h:32]: The code in the cc file dereferences the pointer without any nulllpointer checks, so if you passed nullptr here you would segfault. Either we can never pass nullptr to this class (in which case taking a const& is recommended) or we have to add null checks to all the functions. I prefer the const& approach since it seems nonsensical to delegate to a null route (just pass the null route)

For lifetime I think its fine to copy around a shared ptr if you already have one, though if you want to pass a raw pointer/ref around we need to document the expected lifetime of it, since nothing in code prevents that from being deallocated somewhere else while we're still holding onto the pointer/ref. This could be as easy as saying "X must outlive Y" in comments, though we'll want to make sure that's actually the case

File: docs/root/intro/life_of_a_request.rst
Comment (mengmichael1) [reply docs/root/intro/life_of_a_request.rst:437]: Yeah `tools/code_format/check_format.py fix` didn't work for this RST file, it's ok though I was able to wrap the lines manually!

File: include/envoy/http/filter.h
Comment (mengmichael1) [reply include/envoy/http/filter.h:223]: Thanks for raising. This comment also triggered me to consider if we wanted this to be a reference parameter `const Router::RouteConstSharedPtr&` instead. I saw existing usages ([http/utility](https://github.com/envoyproxy/envoy/blob/main/source/common/http/utility.cc#L916-L943) and [ext_authz filter](https://github.com/envoyproxy/envoy/blob/main/source/extensions/filters/http/ext_authz/ext_authz.cc#L378-L392), among others) of RouteConstSharedPtr being a reference param, and conducting the same `route->routeEntry()` calls.

After giving it some thought and consulting https://docs.microsoft.com/en-us/cpp/cpp/how-to-create-and-use-shared-ptr-instances?view=msvc-160#example-5, I think its best to keep it as is, passing the shared_ptr by value. Since we want the callee `setRoute` to be an owner, and we're not too concerned with the copy operation performance. Do you agree?

File: source/common/router/delegating_route_impl.h
Comment (mengmichael1) [reply source/common/router/delegating_route_impl.h:18]: Related to your comment here (https://github.com/envoyproxy/envoy/pull/15266/files#r586565475) about dereferencing the pointer without nullptr checks, I notice I do that with this `DelegatingRoute` class as well (oops).

I plan on adding in nullptr checks, since I think the cleaner implementation is to pass `RouteConstSharedPtr` by value instead of reference. For example, then callers (filter callbacks) don't need to worry about passing in a reference or checking if the `decoder_callbacks_->route()` is nullptr.

Should `DelegatingRoute` propagate nullptr is nullptr is detected? Ex: `base_route_` is nullptr, then all methods such as `routeEntry()` and `tracingConfig` also return nullptr

File: source/common/router/delegating_route_impl.h
Comment (mengmichael1) [reply source/common/router/delegating_route_impl.h:32]: Ah, I see. Yeah I agree `const Router::RouteEntry&` is the right decision here, since it is nonsensical to delegate to a null route. 

Re: lifetime, got it. We could document something like `@param route_entry must outlive the lifetime of this DelegatingRouteEntry class`. I don't think it'd get deallocated somewhere else, but is there a way to really make sure? Would it be smarter to play it safe and have it be a ConstSharedPtr like the `DelegatingRoute` constructor? But then we'd run into the original issue of needing to add null checks to all the functions.

File: source/common/router/delegating_route_impl.h
Comment (snowp) [reply source/common/router/delegating_route_impl.h:18]: >For example, then callers (filter callbacks) don't need to worry about passing in a reference or checking if the decoder_callbacks_->route() is nullptr.
route() can already return nullptr so callers already handle it. Because of this I think it makes more sense to avoid wrapping a nullptr and have the delegating route assume that its wrapping a non-null route. If we need to pass
a shared ptr to have the delegated route participate in ownership we can `ASSERT(route_ != nullptr)` in the ctor
and skip all the null checks. The suggestion to use const ref over raw-pointer makes sense if we're dealing with
a non-owning pointer, but if we need a shared pointer then an ASSERT is the best we can do.

File: include/envoy/http/filter.h
Comment (mengmichael1) [reply include/envoy/http/filter.h:223]: It’s the same as when `snapped_route_config_->route(cb, *request_headers_, filter_manager_.streamInfo(), stream_id_);` returns nullptr when no route match is found, since I essentially grabbed the core logic from the [current refreshCachedRoute implementation](https://github.com/envoyproxy/envoy/blob/main/source/common/http/conn_manager_impl.cc#L1252-L1279). So the nullptr case is already accounted for.  

 `clearRouteCache` is slightly different in that it sets the cached attributes to optional ptrs 
  ``` 
  cached_route_ = absl::optional<Router::RouteConstSharedPtr>();
cached_cluster_info_ = absl::optional<Upstream::ClusterInfoConstSharedPtr>();
``` 

Would my `setRoute` API make more sense if I replicate the `clearRouteCache` behavior (setting to optional pointers) when route is nullptr?

Comment (mengmichael1) [reply include/envoy/http/filter.h:223]: Apologies, for some reason, Github doesn't resolve my cpp Markdown code block correctly. But hopefully you're able to parse my question :D

File: source/common/router/delegating_route_impl.h
Comment (mengmichael1) [reply source/common/router/delegating_route_impl.h:18]: Thanks for the explanation, makes a lot of sense. I added an `ASSERT(base_route_ != nullptr);` to the DelegatingRoute ctor so it too can participate in ownership (will push up my branch commits soon).

File: source/common/router/delegating_route_impl.h
Comment (mengmichael1) [reply source/common/router/delegating_route_impl.h]: Actually, after chatting with @theevocater about this, we think it could be a good idea to change the "base" that `DelegatingRouteEntry` uses for delegation from a `const Router::RouteEntry&` to a `Router::RouteConstSharedPtr` (done in commit https://github.com/envoyproxy/envoy/pull/15266/commits/27c42ced9e4bd690105845de543f61155e7f110d) 

The ctor is cleaner, we don't need to worry about the raw pointer/ref lifetime of RouteEntry, and `RouteConstSharedPtr` was meant to be shared. What do you think?

File: test/integration/protocol_integration_test.cc
Comment (mengmichael1) [reply test/integration/protocol_integration_test.cc:148]: It's not, but I was hoping to use it since the first use case of this setRoute / DelegatingRoute functionality will be for the Lyft context propagation filter, which will re-route requests to an ORIGINAL_DST cluster

File: source/common/http/filter_manager.cc
Comment (snowp) [inline source/common/http/filter_manager.cc:246]: `std::move(route)` to avoid a copy

File: include/envoy/http/filter.h
Comment (snowp) [inline include/envoy/http/filter.h:235]: Let's document what the implication of passing nullptr here is

File: source/common/router/delegating_route_impl.h
Comment (snowp) [inline source/common/router/delegating_route_impl.h:94]: `const`

File: source/common/router/delegating_route_impl.h
Comment (snowp) [inline source/common/router/delegating_route_impl.h:41]: `std::move(route)`

File: test/common/http/conn_manager_impl_test_base.h
Comment (snowp) [inline test/common/http/conn_manager_impl_test_base.h:222]: `std::move(base_route)`

File: test/common/http/conn_manager_impl_test_base.h
Comment (snowp) [inline test/common/http/conn_manager_impl_test_base.h:235]: `std::move(base_route)`

File: test/integration/filters/set_route_filter.cc
Comment (snowp) [inline test/integration/filters/set_route_filter.cc:23]: `auto` to avoid repeating the type

File: test/integration/filters/set_route_filter.cc
Comment (snowp) [inline test/integration/filters/set_route_filter.cc:31]: would be easy enough to add this to a file in /test/common/http to be reusable, maybe do that in this PR?

File: test/integration/protocol_integration_test.cc
Comment (snowp) [inline test/integration/protocol_integration_test.cc:172]: If we're gonna keep this to test the original dst cluster we should at least comment why we need all this additional logic in this test

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks overall this looks good, just a few comments + nits on std::move usage

File: test/integration/filters/set_route_filter.cc
Comment (mengmichael1) [reply test/integration/filters/set_route_filter.cc]: done

**Review by mattklein123 — COMMENTED** [review]
Thanks at a high level this looks great! I will defer to @snowp for further review/merge.

**Comment by mengmichael1** [comment]
@snowp re: setRoute(nullptr)

I realized that `ConnectionManagerImpl::ActiveStream::route(const Router::RouteCallback& cb)` uses `if (cached_route_.has_value())`, rather than `bool hasCachedRoute() { return cached_route_.has_value() && cached_route_.value(); }` helper, to check whether or not to `refreshCachedRoute`. 

As a result, when the `cached_route` is nullptr, `cached_route_.has_value()` is true and `nullptr` gets returned directly to the caller, no route resolution/refreshing takes place. 

Was this an intentional decision? Perhaps because nullptr indicates route resolution found no suitable route match?

 Asking because, in order for `setRoute(nullptr)` to reach parity with `clearRouteCache()`, I’ll need to either: 
(1) update the conditional from `if (cached_route_.has_value())` to `if hasCachedRoute()` in https://github.com/envoyproxy/envoy/blob/main/source/common/http/conn_manager_impl.cc#L1589-L1595

or

 (2) have a conditional within my `setRoute` such that if nullptr is passed in, I do a `clearRouteCache()`, which sets `cached_route_` and `cached_cluster_info_` to `absl::optional` ptrs instead of nullptrs.
 
 What do you think?

File: test/integration/filters/set_route_filter.cc
Comment (mengmichael1) [reply test/integration/filters/set_route_filter.cc]: done

File: source/common/http/filter_manager.cc
Comment (mengmichael1) [reply source/common/http/filter_manager.cc]: yup done, thanks for pointing these out! 

https://github.com/envoyproxy/envoy/pull/15266/commits/54afe69c91609d8365f22c73b74b1e7c40e4ba22

File: source/common/router/delegating_route_impl.h
Comment (mengmichael1) [reply source/common/router/delegating_route_impl.h]: done

**Comment by snowp** [comment]
I think it might make more sense to have setRoute(nullptr) set the resolved route to nullptr, as if we attempted route resolution and failed to find a route. I don't feel that strongly about what the behavior should be when we pass nullptr, only that we can document the semantics of it and that we're not changing existing behavior in order to incorporate it.

File: include/envoy/http/filter.h
Comment (mengmichael1) [reply include/envoy/http/filter.h]: resolving, will differentiate setRoute(nullptr) from clearRouteCache() per this discussion https://github.com/envoyproxy/envoy/pull/15266#issuecomment-808517011

File: include/envoy/http/filter.h
Comment (mengmichael1) [reply include/envoy/http/filter.h:244]: done, see https://github.com/envoyproxy/envoy/pull/15266/commits/9012f546faaa951334f11ac18d76406b62d1bcae

File: test/common/http/conn_manager_impl_test.cc
Comment (mengmichael1) [inline test/common/http/conn_manager_impl_test.cc:790]: Determined from https://github.com/envoyproxy/envoy/blob/effa486d042cd2ac141ee48d8ed54b02b0951c73/source/common/router/router.cc#L409-L410

**Comment by mengmichael1** [comment]
Hey @snowp, this is ready for another pass :) 

As discussed offline, I moved my integration test out of ProtocolIntegration and into a separate standalone test (see `set_delegating_route_test.cc`)

**Review by snowp — CHANGES_REQUESTED** [review]
Thanks this looks great! 

@alyssawilk for a non-Lyft review of the HCM changes

**Comment by alyssawilk** [comment]
I'm headed out (EoD EST) and out tomorrow but I'll check it out early Monday!

File: source/common/router/delegating_route_impl.cc
Comment (alyssawilk) [inline source/common/router/delegating_route_impl.cc:52]: I was going to check coverage on this but for some reason the successful CI runs only latched DCO.
Can you make sure we have unit tests for all the delegating functions in this PR?

File: source/common/router/delegating_route_impl.h
Comment (alyssawilk) [inline source/common/router/delegating_route_impl.h:21]: nit: mind adding comment
// Router::Route

File: test/integration/set_delegating_route_test.cc
Comment (alyssawilk) [inline test/integration/set_delegating_route_test.cc:8]: optional nit, I think this would be fine folded into another integration test (intergration_test.cc would be fine, or http2_integration_test.cc if you want to keep it H2) so we have one less thing to link.)

File: test/integration/set_delegating_route_test.cc
Comment (alyssawilk) [inline test/integration/set_delegating_route_test.cc:62]: Network::Test::getLoopbackAddressString(GetParam()) ?

File: test/integration/set_delegating_route_test.cc
Comment (alyssawilk) [inline test/integration/set_delegating_route_test.cc:65]: oh, or I think you can just use
fake_upstreams_[1]->localAddress().asString() and skip the loopback helper entirely.

File: test/integration/set_delegating_route_test.cc
Comment (alyssawilk) [inline test/integration/set_delegating_route_test.cc:80]: This is a really good explanation, thanks.
I'd also encourage useAccessLog("%UPSTREAM_CLUSTER%")  / waitForAccessLog() and checking name, as I feel like it's a bit less fragile.

**Review by alyssawilk — COMMENTED** [review]
Looks great!
I'd encourage adding this into the release notes as a new feature as I think it's useful enough we should call it out :-)

**Comment by alyssawilk** [comment]
Oh also looks like CI is wedged -a new push should sort that out.

File: test/integration/set_delegating_route_test.cc
Comment (mengmichael1) [reply test/integration/set_delegating_route_test.cc]: Love the suggestion, definitely makes the test less fragile. Was looking for something just like this!

File: test/integration/set_delegating_route_test.cc
Comment (mengmichael1) [reply test/integration/set_delegating_route_test.cc]: One less thing to link is nice, done

File: source/common/router/delegating_route_impl.cc
Comment (mengmichael1) [reply source/common/router/delegating_route_impl.cc:52]: I added a new unit test `DelegatingRouteEntryAllCalls`  to add coverage for all routeEntry delegating methods (except typedMetadata, ran into a strange issue with its mock not having defaults set, added a comment about it).

  Is this what you were imagining with "unit tests for all the delegating functions"?
  
    Left other Router::Route stuff alone (e.g. `directResponseEntry`) since this PR doesn’t define a "DelegatingDirectResponseEntry" mechanism

**Comment by mengmichael1** [comment]
Thanks for the review @alyssawilk! I think this is ready for another pass now. You mentioned adding this to the release notes, how do I do that?

**Review by alyssawilk — COMMENTED** [review]
LGTM modulo ci being sorted out and release notes (glad you found the right file, and the other additions are fantastic!)

Also as I'm out tomorrow feel free to ping snow for merge once those two are addressed!

**Review by alyssawilk — DISMISSED** [review]
LGTM!  Will defer merge to snow since I'm headed out for a long weekend shortly :-)

**Comment by snowp** [comment]
@envoyproxy/windows-dev There seems to be some Windows only (I think tsan is unrelated? but not 100% sure) on this PR, any chance anyone of you could help debugging? The error message is unfortunately not very helpful

**Comment by davinci26** [comment]
I have some time now, I can take a look

File: test/integration/integration_test.cc
Comment (davinci26) [inline test/integration/integration_test.cc:2017]: if you use `Network::Test::getLoopbackAddressUrlString` instead of `Network::Test::getLoopbackAddressString` `getLoopbackAddressUrlString` it should pass on windows

Comment (mengmichael1) [reply test/integration/integration_test.cc]: thanks for the pointer! you are right, I'm missing [the brackets needed](https://github.com/envoyproxy/envoy/blob/main/test/test_common/network_utility.cc#L74-L79) for IPv6 

trying out now (92f250d)

**Review by snowp — APPROVED** [review]
LGTM, thanks! Happy to see this land!

**Pull request merged**
